src/server/scripts/Northrend/Naxxramas/boss_thaddius.cpp

```diff
@@ -110,168 +155,270 @@ class boss_thaddius : public CreatureScript
+                if (_phase == PHASE_TRANSITION || _phase == PHASE_THADDIUS)
+                    BeginResetEncounter();
             }
 
-            polaritySwitch = false;
-            uiAddsTimer = 0;
-        }
-
-        bool checkStalaggAlive;
-        bool checkFeugenAlive;
-        bool polaritySwitch;
-        uint32 uiAddsTimer;
-
-        void KilledUnit(Unit* /*victim*/) override
-        {
-            if (!(rand32() % 5))
-                Talk(SAY_SLAY);
-        }
-
-        void JustDied(Unit* /*killer*/) override
-        {
-            _JustDied();
-            Talk(SAY_DEATH);
-        }
+            void JustDied(Unit* /*killer*/) override
+            {
+                _JustDied();
+                me->setActive(false);
+                if (Creature* stalagg = ObjectAccessor::GetCreature(*me, instance->GetGuidData(DATA_STALAGG)))
+                    stalagg->setActive(false);
+                if (Creature* feugen = ObjectAccessor::GetCreature(*me, instance->GetGuidData(DATA_FEUGEN)))
+                    feugen->setActive(false);
+                Talk(SAY_DEATH);
+            }
 
-        void DoAction(int32 action) override
-        {
-            switch (action)
+            void DoAction(int32 action) override
             {
-                case ACTION_FEUGEN_RESET:
-                    checkFeugenAlive = true;
-                    break;
-                case ACTION_FEUGEN_DIED:
-                    checkFeugenAlive = false;
-                    break;
-                case ACTION_STALAGG_RESET:
-                    checkStalaggAlive = true;
-                    break;
-                case ACTION_STALAGG_DIED:
-                    checkStalaggAlive = false;
-                    break;
+                switch (action)
+                {
+                    case ACTION_RESET_ENCOUNTER_TIMER:
+                        if (_phase == PHASE_RESETTING)
+                            ResetEncounter();
+                        break;
+                    case ACTION_FEUGEN_RESET:
+                    case ACTION_STALAGG_RESET:
+                        if (_phase != PHASE_NOT_ENGAGED && _phase != PHASE_RESETTING)
+                            BeginResetEncounter();
+                        break;
+                    case ACTION_FEUGEN_AGGRO:
+                    case ACTION_STALAGG_AGGRO:
+                        if (_phase == PHASE_RESETTING)
+                            return BeginResetEncounter();
+                        if (_phase != PHASE_NOT_ENGAGED)
+                            return;
+                        _phase = PHASE_PETS;
+                        events.SetPhase(PHASE_PETS);
+
+                        shockingEligibility = true;
+                        
+                        if (!instance->CheckRequiredBosses(BOSS_THADDIUS))
+                            return BeginResetEncounter();
+                        instance->SetBossState(BOSS_THADDIUS, IN_PROGRESS);
+
+                        me->setActive(true);
```

Comment:
Just trying to prevent weirdness if the last person to drop from combat is also the last person that hasn't released and leaves grid just they exit combat (hearthstone etc.)

BossAI does it using the default _EnterCombat, too, which is where I got the idea.